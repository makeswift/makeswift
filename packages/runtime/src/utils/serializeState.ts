import { mapValues } from '@makeswift/controls'

export type SerializedState<T> =
  T extends Map<string, infer U>
    ? Record<string, SerializedState<U>>
    : T extends Set<infer U>
      ? Array<SerializedState<U>>
      : T extends Record<string, infer U>
        ? Record<string, SerializedState<U>>
        : T extends Array<infer U>
          ? Array<SerializedState<U>>
          : T

export function serializeState<T>(state: T): SerializedState<T> {
  if (state instanceof Map) {
    return serializeState(Object.fromEntries(state.entries())) as SerializedState<T>
  }

  if (state instanceof Set) return serializeState([...state.values()]) as SerializedState<T>

  if (Array.isArray(state)) return state.map(serializeState) as SerializedState<T>

  if (state !== null && typeof state === 'object') {
    return mapValues(state, serializeState) as SerializedState<T>
  }

  return state as SerializedState<T>
}
