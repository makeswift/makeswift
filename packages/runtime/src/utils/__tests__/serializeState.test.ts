import { expectTypeOf } from 'expect-type'
import { serializeState } from '../serializeState'

describe('serializeState', () => {
  test.each([17, 'text', ['hello', 42.05, 'world']])('correctly serializes %s to itself', state => {
    expect(serializeState(state)).toStrictEqual(state)
  })

  test.each([
    new Map<string, any>([
      ['text', 'hi'],
      ['number', 24.99],
      ['array', ['hello', 17]],
      [
        'record',
        {
          big: 'fish',
          small: 'pond',
          nestedArray: [3, 5, 7, 9],
          nestedMap: new Map([['key', 'value']]),
          nestedSet: new Set([1, 2, 3]),
        },
      ],
      ['set', new Set([3, 5, 7])],
      ['map', new Map([['price', 9.99]])],
    ]),
  ])('correctly serializes %s', state => {
    expect(serializeState(state)).toMatchSnapshot()
  })

  test('correctly infers the result type', () => {
    expectTypeOf(serializeState(new Set([3, 5, 7]))).toEqualTypeOf<number[]>()

    expectTypeOf(serializeState(new Map([['text', 'hi']]))).toEqualTypeOf<Record<string, string>>()
    expectTypeOf(serializeState(new Map([['nested', new Map([['number', 17]])]]))).toEqualTypeOf<
      Record<string, Record<string, number>>
    >()

    expectTypeOf(
      serializeState({ record: new Map([['nested', new Map([['number', 17]])]]) }),
    ).toEqualTypeOf<Record<string, Record<string, Record<string, number>>>>()

    expectTypeOf(
      serializeState({ record: new Map([['nested', new Set([3, 5, 7])]]) }),
    ).toEqualTypeOf<Record<string, Record<string, number[]>>>()
  })
})
