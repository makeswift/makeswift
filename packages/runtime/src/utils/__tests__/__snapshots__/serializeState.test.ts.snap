// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeState correctly serializes Map(6) {
  'text' => 'hi',
  'number' => 24.99,
  'array' => [Array],
  'record' => [Object],
  'set' => [Set],
  'map' => [Map]
} 1`] = `
{
  "array": [
    "hello",
    17,
  ],
  "map": {
    "price": 9.99,
  },
  "number": 24.99,
  "record": {
    "big": "fish",
    "nestedArray": [
      3,
      5,
      7,
      9,
    ],
    "nestedMap": {
      "key": "value",
    },
    "nestedSet": [
      1,
      2,
      3,
    ],
    "small": "pond",
  },
  "set": [
    3,
    5,
    7,
  ],
  "text": "hi",
}
`;
